pkg/hooks/manager.go

```diff
@@ -0,0 +1,228 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package hooks
+
+import (
+	"context"
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	hooksInfo "kubevirt.io/kubevirt/pkg/hooks/info"
+	hooksV1alpha1 "kubevirt.io/kubevirt/pkg/hooks/v1alpha1"
+	"kubevirt.io/kubevirt/pkg/log"
+	domainSchema "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	virtwrapApi "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+)
+
+type callackClient struct {
+	SocketPath          string
+	Version             string
+	subsribedHookPoints []*hooksInfo.HookPoint
+}
+
+var manager *Manager
+var once sync.Once
+
+type Manager struct {
+	callbacksPerHookPoint map[string][]*callackClient
+}
+
+func GetManager() *Manager {
+	once.Do(func() {
+		manager = &Manager{callbacksPerHookPoint: make(map[string][]*callackClient)}
```

Comment:
@rmohr I could use another set of eyes on this global manager variable.

I *think* we're safe with respects to concurrency because the callbacks map is initialized in once before any goroutines attempt to read it. after initialization no goroutine attempts to write again, so it's essentially read only.

can you verify my sanity that no locks are necessary here?

Reply:
Oh sorry, did not realize that from your previous comment, I think you are right here. Let's wait for Roman.

Reply:
Yes it should work without a lock, especially since later invocations don't change the value anymore.
Do we need a singleton here? Can't we just create the manager once and pass it  to the libvirt manager  on creation?

Reply:
I will remove the lock. I made it a singleton so it is easier to use it from different places (not sure where it will be used) without the need to pass the instance everywhere.

Reply:
I think the do.once part is fine. There shouldn't be anything to change here. 

It's given everywhere we'll potentially have to pass this object, it's going to be easier to keep it as a singleton.

Reply:
We have other objects like informers which are also singletons, because we want to be sure that they are only used once. Still they are injected into structs on controller creation time. I think it is good that it is a singleton, if it has to be one (technical wise) but I would still prefer if we inject it into a struct field of the manager during creation.

Reply:
Which manager? LibvirtDomainManager? Aren't we going to create also non-domain hook points?

Reply:
If we do that I would expect that other places inject the singleton from factory-code and not directly call the method. It just makes refactoring and testing easier. We  do the same for other singletons in the project. For instance if you then mock the hook, it is not clear in other tests what you get. Do you get the real hook manager? Do you get the mock from the laste test?